cloudapp/src/app/excel-export/excel-export-service.ts

```diff
@@ -0,0 +1,137 @@
+
+import * as XLSX from 'sheetjs-style'
+import * as FileSaver from 'file-saver'
+import { RequestedResource, RequestedResourcesService } from '../requested-resources-service/requested-resources-service'
+
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ExcelExportService {
+  private static readonly FILE_NAME: string = 'requested-resources'
+  private static readonly EXCEL_TYPE: string = 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;charset=UTF-8'
+  private static readonly EXCEL_EXTENSION: string = '.xlsx'
+
+  private static readonly PAGE_SIZE: number = 100
+
+  constructor(
+    private requestedResourcesService: RequestedResourcesService
+  ) { }
+
+
+  async generateExcel(circDeskCode: string, libraryCode: string, columnOptions: ColumnOption[]) {
+
+    const columnDefinitions: ColumnDefinition[] = this.getColumnDefinitions(columnOptions)
+
+    this.requestedResourcesService.findRequestedResources(
+		circDeskCode, 
+		libraryCode, 
+		ExcelExportService.PAGE_SIZE, 
+		null, 
+		(count:number) => { console.log('completed', count) }, 
+		(err:Error) => { console.log('error', err) })
+      .then(resources => {
+        const data: string[][] = this.createOutputFormat(resources, columnDefinitions)
+        const worksheet: XLSX.WorkSheet = XLSX.utils.aoa_to_sheet(data)
+        this.setCellWith(worksheet, columnDefinitions)
+        this.setCellStyles(worksheet)
+        const workbook: XLSX.WorkBook = { Sheets: { 'data': worksheet }, SheetNames: ['data'] }
+        const excelBuffer: any = XLSX.write(workbook, { bookType: 'xlsx', type: 'array' })
+        this.saveAsExcelFile(excelBuffer)
+      })
+      .catch(e => console.log(e))
+  }
+
+
+  private getColumnDefinitions(columnOptions: ColumnOption[]): ColumnDefinition[] {
+    let columnDefinitions: ColumnDefinition[] = []
+    columnOptions.map(option => {
+      columnDefinitions.push(COLUMNS_DEFINITIONS.get(option.code))
+    })
+    return columnDefinitions
+  }
+
+
+  private createOutputFormat(resources: RequestedResource[], columnDefinitions: ColumnDefinition[]): string[][] {
+    let data: string[][] = resources.map(resource => this.mapColumns(resource, columnDefinitions))
+    data.unshift(this.createHeader(columnDefinitions))
+    return data
+  }
+
+
+  private mapColumns(requestedResource: RequestedResource, columnDefinitions: ColumnDefinition[]): string[] {
+    return columnDefinitions.map(col => {
+      try {
+        let value:string = COLUMNS_DEFINITIONS.get(col.code).mapFn(requestedResource)
+        if(!value) {
+          value = '-'
+        }
+        return value
+      } catch (err) {
+        console.error(`Failed to mapped column ${col.name} for `, requestedResource, err)
+        return undefined
+      }
+    })
+  }
+
+
+  private setCellWith(worksheet: XLSX.WorkSheet, columnDefinitions: ColumnDefinition[]) {
+    //const colsWiths: object[] = columnDefinitions.map(col => { return { wch: col.width } })
```

Comment:
This is interesting. Does it make sense for the width column option to be the Excel column width, and not the maximum number of characters?

Reply:
During testing, I added preset column widths to the ColumnDefinitions, since this is not the correct place to add information like this, I removed it. 
The width should be configurable in the ColumnOptions, but this is an open todo.

At the moment, I removed the comment in commit b075a3c